cargo/workspace.bzl

```diff
@@ -0,0 +1,251 @@
+        "--input_path",
+        input_path,
+        "--output_path",
+        repository_ctx.path("defs.bzl"),
+        "--repo-name",
+        repository_ctx.attr.name,
+    ]
+    if lockfile_path != None:
+        args.append("--lockfile")
+        str(args.append(lockfile_path))
+    if repository_ctx.os.environ.get("RULES_RUST_UPDATE_CRATE_UNIVERSE_LOCKFILE", "false") == "true":
+        args.append("--update-lockfile")
+
+    path_binaries = [
+        repository_ctx.path(rustc_label),
+    ]
+
+    result = repository_ctx.execute(
+        args,
+        environment = {
+            "RUST_LOG": "info",
+
+            # The resolver invokes `cargo metadata` which relies on `rustc` being on the $PATH
+            "PATH": ":".join([str(b.dirname) for b in path_binaries]),
+        },
+    )
+    repository_ctx.delete(input_path)
+    if result.stderr:
+        print("Output from resolver: " + result.stderr)
+    if result.return_code != 0:
+        fail("Error resolving deps:\n" + result.stdout + "\n" + result.stderr)
+
+    repository_ctx.file("BUILD.bazel")
+
+_crate_universe_resolve = repository_rule(
+    implementation = _crate_universe_resolve_impl,
+    attrs = {
+        "packages": attr.string_list(allow_empty = True),
+        "cargo_toml_files": attr.label_list(),
+        "overrides": attr.string_dict(doc = """Mapping of crate name to crate.override(...) entries)
+
+Example:
+
+    load("@rules_rust//cargo:workspace.bzl", "rust_library", "crate")
+
+    rust_library(
+        name = "mylib",
+        # [...]
+        overrides = {
+            "tokio": crate.override(
+                extra_rust_env_vars = {
+                    "MY_ENV_VAR": "MY_ENV_VALUE",
+                },
+                extra_build_script_env_vars = {
+                    "MY_BUILD_SCRIPT_ENV_VAR": "MY_ENV_VALUE",
+                },
+                extra_bazel_deps = {
+                    # Extra dependencies are per target. They are additive.
+                    "cfg(unix)": ["@somerepo//:foo"],  # cfg() predicate.
+                    "x86_64-apple-darwin": ["@somerepo//:bar"],  # Specific triple.
+                    "cfg(all())": ["@somerepo//:baz"],  # Applies to all targets ("regular dependency").
+                },
+                extra_build_script_bazel_deps = {
+                    # Extra dependencies are per target. They are additive.
+                    "cfg(unix)": ["@buildscriptdep//:foo"],
+                    "x86_64-apple-darwin": ["@buildscriptdep//:bar"],
+                    "cfg(all())": ["@buildscriptdep//:baz"],
+                },
+                extra_bazel_data_deps = {
+                    # ...
+                },
+                extra_build_script_bazel_data_deps = {
+                    # ...
+                },
+            ),
+        },
+    )
+
+"""),
+        "repository_template": attr.string(),
+        "supported_targets": attr.string_list(allow_empty = False),
+        "lockfile": attr.label(
```

Comment:
As it's written now, this is misleading and should either be renamed or have a doc explaining that it's not actually a `Cargo.lock` file.

Reply:
Do you have a better name? Lockfile is the general term of art for what this is, and its name contrasts with "cargo_toml_files" - I'd name it "cargo_lock_file" were it a Cargo.lock...

Reply:
I think there's two much mixed functionality setting different expectations in this initial approach. My primary use case is to be able to take a `Cargo.toml` file and a `Cargo.lock` file and generate all the same dependencies Cargo would. But that doesn't seem to be the case here, unless I missed something. 

I also struggle to think of why someone would mix and match `cargo_toml_files` and `packages`. It seems that if you're already using `Cargo.toml` files, it's not an unfamiliar process to go update or make a new one to get more dependencies. Only in a case where I have a pure Bazel repo would I directly add `packages`. I suspect having both means you can no longer trust the `Cargo.lock` file because you're constantly dealing with a modified `Cargo.toml` file internally.

(Controversial statement inbound 😅) I think `lockfile` is fine, but I'd say instead of `allow_single_file = True`, it'd be `allow_single_file = [".lock", ".bzl"]` and the use of `Cargo.toml` files be restricted from mixing with directly named `packages` so it's clear where what type of file should be used (unless you want to have multiple attributes but I think it's fine to have a single `lockfile` attribute).